cmd/juju/caas/remove.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package caas
+
+import (
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"github.com/juju/gnuflag"
+	"github.com/juju/juju/cloud"
+
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/api/base"
+	cloudapi "github.com/juju/juju/api/cloud"
+	"github.com/juju/juju/cmd/modelcmd"
+	"github.com/juju/juju/jujuclient"
+)
+
+var usageRemoveCAASSummary = `
+Removes a k8s endpoint from Juju.`[1:]
+
+var usageRemoveCAASDetails = `
+Removes the specified k8s cloud from the controller and user-defined
+cloud details (if it is not in use).
+
+Examples:
+    juju remove-k8s myk8scloud
+    
+See also:
+    add-k8s
+`
+
+// Implemented by cloudapi.Client
+type RemoveCloudAPI interface {
+	RemoveCloud(string) error
+	Close() error
+}
+
+// RemoveCAASCommand is the command that allows you to remove a k8s cloud.
+type RemoveCAASCommand struct {
+	modelcmd.ControllerCommandBase
+
+	// cloudName is the name of the caas cloud to remove.
+	cloudName string
+
+	cloudMetadataStore  CloudMetadataStore
+	fileCredentialStore jujuclient.CredentialStore
+	apiRoot             api.Connection
```

Comment:
I am not sure you actually want to cache this, rather than have a function variable to obtain a new one.

Since you are defering a .Close() on it in your Run() method, it is possible to imagine a scenario where an api is closed and is not nil but will be returned by current newAPIRoot().

Reply:
This is just the pattern used elsewhere as well. Commands are only ever instantiated once. They get the connection to the controller once. It's an attribute on the command so that tests can pass in a no-op one. Sadly we're probably not consistent - I think other commands pass in a getter func. So we have a mix of approaches. I used this approach here to be consistent with other existing commands in the neighbourhood. I guess we could change them all to use a helper func.

Reply:
Please do :D At least, fix it here.

We have been caught by that pattern before and that is precisely why the func approach has evolved.